http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/ProxyTunnelInitHandlerTest.java

```diff
@@ -0,0 +1,193 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.DefaultChannelPromise;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.util.concurrent.Promise;
+import java.io.IOException;
+import java.net.URI;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+/**
+ * Unit tests for {@link ProxyTunnelInitHandler}.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ProxyTunnelInitHandlerTest {
+    private static final NioEventLoopGroup GROUP = new NioEventLoopGroup(1);
+
+    private static final URI REMOTE_HOST = URI.create("https://s3.amazonaws.com:1234");
+
+    @Mock
+    private ChannelHandlerContext mockCtx;
+
+    @Mock
+    private Channel mockChannel;
+
+    @Mock
+    private ChannelPipeline mockPipeline;
+
+    @Mock
+    private ChannelPool mockChannelPool;
+
+    @Before
+    public void methodSetup() {
+        when(mockCtx.channel()).thenReturn(mockChannel);
+        when(mockCtx.pipeline()).thenReturn(mockPipeline);
+        when(mockChannel.pipeline()).thenReturn(mockPipeline);
+        when(mockChannel.writeAndFlush(anyObject())).thenReturn(new DefaultChannelPromise(mockChannel, GROUP.next()));
+    }
+
+    @AfterClass
+    public static void teardown() {
+        GROUP.shutdownGracefully().awaitUninterruptibly();
+    }
+
+    @Test
+    public void addedToPipeline_addsCodec() {
+        ProxyTunnelInitHandler handler = new ProxyTunnelInitHandler(mockChannelPool, REMOTE_HOST, null);
+        handler.handlerAdded(mockCtx);
+        verify(mockPipeline).addBefore(anyString(), eq(null), any(HttpClientCodec.class));
```

Comment:
This seemed like a code smell, but then I checked the implementation and saw that we were always passing in a new HttpClientCodec. Should we be using a factory for this and better control it in testing?